src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id, 'version' => $version->toInt()]);
+
+        $row = $stmt->fetch(FetchMode::ASSOCIATIVE);
+
+        if (!$row) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $row['version']) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        return $this->rowToItemVersion($row);
+    }
+
+    public function list(int $limit = 10, ?string $cursor = null) : ItemListPage
+    {
+        if (null !== $cursor && ((string) (int) $cursor) !== $cursor) {
+            return new ItemListPage([], null);
+        }
+
+        $query = $this->connection->createQueryBuilder();
+
+        $query
+            ->addSelect('item.id')
+            ->addSelect('item.sequence')
+            ->from($this->tableItems, 'item')
+            ->orderBy('item.sequence')
+            ->setMaxResults($limit + 1);
+
+        if ($cursor) {
+            $query
+                ->where($query->expr()->gte('item.sequence', ':cursor'))
+                ->setParameter('cursor', (int) $cursor);
+        }
+
+        $results = $query->execute();
+        $ids = $results->fetchAll(FetchMode::ASSOCIATIVE);
+
+        if (empty($ids)) {
+            return new ItemListPage([], null);
+        }
+
+        if (count($ids) > $limit) {
+            $newCursor = (string) array_pop($ids)['sequence'];
+        }
+
+        $ids = array_map([ItemId::class, 'fromString'], array_column($ids, 'id'));
+
+        return new ItemListPage($ids, $newCursor ?? null);
+    }
+
+    public function count() : int
+    {
+        $results = $this->connection->executeQuery("SELECT COUNT(item.id) count FROM {$this->tableItems} item");
+
+        return (int) $results->fetch()['count'];
+    }
+
+    public function getIterator() : Traversable
+    {
+        while (true) {
+            /** @var Statement $statement */
+            $results = $this->connection->executeQuery(
+                'SELECT version.id, version.version, version.content, version.hash, item.sequence'
+                ." FROM {$this->tableItems} item LEFT JOIN 
+{$this->tableVersions} version ON item.id = version.id AND item.version = version.version AND item.sequence > :sequence"
+                .' ORDER BY item.sequence LIMIT 100',
+                ['sequence' => $cursor ?? 0]
+            );
+
+            while ($result = $results->fetch(FetchMode::ASSOCIATIVE)) {
+                yield $this->rowToItemVersion($result);
+                $cursor = $result['sequence'];
+            }
+
+            if (empty($result)) {
+                break;
+            }
+        }
+    }
+
+    public function getSchema() : Schema
+    {
+        $schema = new Schema();
+
+        $table = $schema->createTable($this->tableItems);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('sequence', Type::INTEGER, ['autoincrement' => true]);
```

Comment:
it's early to talk about ordering strategies and dates, so a integer sequence guarantees a global ordering :+1: It does become a bottleneck for creation of first versions of items, but seems unlikely to be relevant in practice.